### PR TITLE
hpx: overhaul of the package recipe

### PR DESCRIPTION
**Pull request by alalazo**
Modifications:

- [x] Added latest versions
- [x] Added support for different kind of malloc
- [x] Added support for instrumentation
- [x] Added support to activate or not `cuda`, networking and tools
- [x] Enforced the C++ standard through a variant, make it consistent with `boost`

### CODE REVIEW

**Comment by alalazo** [comment]
@biddisco I am trying to work through the use-case of `hpx`. Is there in your opinion any other configuration handle that should be exposed to users apart from:

- having or not support for networking
- having or not support for cuda
- build or not hpx tools
- instrumentation with apex, perftools, papi and valgrind
- choice of the allocator to be used

?

**Comment by biddisco** [comment]
That covers the most 'obvious' big ones that I can think of. Unfortunately, there are so many options available that one couldn't really do them all. One that is important is
`CMAKE_BUILD_TYPE=Release` or `CMAKE_BUILD_TYPE=Debug` or possibly also `CMAKE_BUILD_TYPE=RelWithDebInfo` because the Release and Debug builds are not compatible with each other and the user's code setting. It's always good to have both builds installed as code will crash if compiled with Release against a Debug HPX (or vice versa).

File: var/spack/repos/builtin/packages/hpx/package.py [hunk 1]
```diff
@@ -7,17 +7,114 @@
...
+
+    variant('instrumentation', values=any_combination_of(
+        'apex', 'google_perftools', 'papi', 'valgrind'
+    ), description='Add support for various kind of instrumentation')
+
+    variant('networking', default=True,
+            description='Support for networking and multi=node runs')
+    variant('tools', default=False, description='Build HPX tools')
+
+    depends_on('boost')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/hpx/package.py]: This line isn't necessary with the line below.

File: var/spack/repos/builtin/packages/hpx/package.py [hunk 2]
```diff
@@ -7,17 +7,114 @@
...
     depends_on('boost@1.55.0:')
     depends_on('hwloc@1.6:')
+    depends_on('python')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/hpx/package.py]: Usually `type=('build', 'run')`?

**Comment by adamjstewart** [comment]
> One that is important is
`CMAKE_BUILD_TYPE=Release` or `CMAKE_BUILD_TYPE=Debug` or possibly also `CMAKE_BUILD_TYPE=RelWithDebInfo`

The CMakePackage base class already comes with a variant for this, so we should be good to go.

File: var/spack/repos/builtin/packages/hpx/package.py (hunk 2 above)
Comment (alalazo) [reply var/spack/repos/builtin/packages/hpx/package.py]: @biddisco I saw that `python` is detected by CMake during the configuration step. Do you know when is it used? Is it just a build tool or do you need it at run-time or link-time?

File: var/spack/repos/builtin/packages/hpx/package.py (hunk 1 above)
Comment (alalazo) [reply var/spack/repos/builtin/packages/hpx/package.py]: Isn't `boost` treated differently from `boost@1.55:` during normalization? I thought adding that would help the concretizer, but thinking twice I'm not sure anymore...

Comment (adamjstewart) [reply var/spack/repos/builtin/packages/hpx/package.py]: Nope. This is handled differently though:
```python
depends_on('boost')
depends_on('boost@1.55:', when='@1.2.3:')
```
Maybe that's what you were thinking of.

File: var/spack/repos/builtin/packages/hpx/package.py (hunk 2 above)
Comment (biddisco) [reply var/spack/repos/builtin/packages/hpx/package.py]: our unit tests are run by a python script called hpxrun.py - each test is added via the usual cmake add_test(blah blah) but the executable is the python script and it launches the actual test and passes params through to it. I don't like it personally, but it was done years ago to support some of the test that launch multiple processes on a node and connect to each other and the python script made it easy to handle.
I don't recall if python is used elsewhere.

Comment (alalazo) [reply var/spack/repos/builtin/packages/hpx/package.py]: @adamjstewart I went with `build`, `test` and `run`. Can be refined later.

**Comment by alalazo** [comment]
@adamjstewart That's ready from my side, if there are no other changes requested.

**Comment by biddisco** [comment]
I had a quick look and just want to point out that https://stellar-group.github.io/hpx/docs/sphinx/latest/html/manual/building_hpx.html#software-and-libraries lists the requirements for dependencies.
I saw boost and hwloc versions in the spack config that look pretty old. Although our minumum requirements are quite lax - we're about to release hpx 1.3 (a few weeks) and we're bumping the requirements up. For hpx 1.2 they are recommended to be

Boost C++ Libraries 1.62.0 (or newer)
Portable Hardware Locality (HWLOC)  1.11 (or newer)

certainly I don' think hwloc 1.6 (in spack file)  should be considered even reasonably close. and we should use these two values as a the new (min) defaults.

**Comment by alalazo** [comment]
@biddisco Thanks for the tip. I'll refine the constraints on both `boost` and `hwloc`.  Notice though that:
```python
depends_on('boost@1.55')
```
is different from:
```python
depends_on('boost@1.55:')  # notice the trailing colon
```
The former means "use boost version 1.55" the latter "use boost version 1.55 or newer".

**Pull request merged**